### DISCUSSION TITLE
yolov5/tutorials/model_pruning_and_sparsity/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#14963] [Docs]
# yolov5/tutorials/model_pruning_and_sparsity/

Learn how to prune YOLOv5 models for improved performance. Follow this step-by-step guide to optimize your YOLOv5 models effectively.

https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity/

**Comment by gauravcodes08** [comment]
anyone tell how to prune and optimize yolov8 models i trained my model on yolov8n for human detection for only one class it reduces the cpu utiliztion but i want to prune the model or sparsity please give me code how to do that and give me others methods as well anyone knows

**Reply by glenn-jocher** [reply]
@Singh0007 to prune and optimize your YOLOv8 model, you can follow similar steps to those used for YOLOv5. Although the provided tutorial is for YOLOv5, the principles remain applicable. First, ensure you are using the latest version of the Ultralytics package. For pruning, you can use the `torch_prune` library to achieve sparsity in your model's weights. Additionally, you might want to explore quantization techniques to further optimize the model. For detailed steps and code examples, please refer to the Ultralytics documentation.

**Reply by hemanthmandava945** [reply]
@glenn-jocher
i was trying prune yolov8x model and i want test the pruned model on my lidar images to know its accuracy and inference speed. below is my code
import torch 
import torch.nn.utils.prune as prune
from ultralytics import YOLO

dataset_path = "/myenv/lib/python3.10/site-packages/ultralytics/cfg/datasets/coco8.yaml"

def prune_model(model,amount = 0.3):
    for module in model.modules():
        if isinstance(module,torch.nn.Conv2d):
            prune.l1_unstructured(module,name="weight", amount=amount)
            prune.remove(module,'weight')
            

        return model
    
model = YOLO("yolov8x-seg.pt")

results  = model.val(data="coco8.yaml")
print(f"map50-95: {results.box.map}")
torch_model = model.model
print(torch_model)
print("pruning model....")
pruned_torch_model = prune_model(torch_model,amount=0.05)
print("model pruned.")
model.model =  pruned_torch_model
print("saving the pruned model...")
model.save("yolov8x_pruned.pt")
print("pruned model saved.")
model = YOLO("yolov8x_pruned.pt")
results = model.val(data = "cocc8.yaml")
print(f"map50-95: {results.box.map}")

i am encountring following error:
  prepared_batch = super()._prepare_batch(si, batch)
  File "/home/mandava/myenv/lib/python3.10/site-packages/ultralytics/models/yolo/detect/val.py", line 107, in _prepare_batch
    cls = batch["cls"][idx].squeeze(-1)
IndexError: The shape of the mask [17] at index 0 does not match the shape of the indexed tensor [0, 1] at index 0
can you help me to solve the issue.

I have few questions on pruning:
1.is it better to do pruning during training or applying pruning after training.as i am researching improving efficiency of yolo model in terms of inference speed. i want to apply pruning and quantization techniques comparing the results.
 2. can you provide me the resources to follow as i could not find many?

**Reply by glenn-jocher** [reply]
For your pruning error, ensure you're applying pruning correctly across all layers and consider fine-tuning after pruning. For YOLOv8 optimization, we recommend using our integrated Neural Magic DeepSparse Engine which handles pruning and quantization efficiently. Check the Ultralytics YOLOv5 Model Pruning tutorial for methodology and explore the Neural Magic integration guide for optimized deployment.
